File: cbse_class_12_txt_to_csv.go
package main

import (
    "bufio"
    "encoding/csv"
    "errors"
    "fmt"
    "regexp"
    "strconv"
    "strings"
    "os"
)

func main() {
    scanner := bufio.NewScanner(os.Stdin)

    for {
        fmt.Println("Enter the (absolute) path to the CBSE Class 12th result file",
         "(should be of the format {SCHOOL_CODE}.TXT).")
        fmt.Println("If this executable program is stored in the same directory, then enter its name only.")
        
        fmt.Print("\nPath: ")
        scanner.Scan()
        path := strings.ToLower(scanner.Text())

        err := writeToCSV(path)
        if err != nil {
            fmt.Println(err)
            fmt.Println("An error occurred. Please try again!\n")
            continue
        } else {
            fmt.Println("Records successfully written to 'class_12th_result.csv'!")
            break
        }
    }
}

func writeToCSV(path string) error {
    if path[len(path) - 4: len(path)] != ".txt" {
        return errors.New("It doesn't looks like a valid result file. Make sure the path is correct.")
    }

    inputFile, err := os.Open(path)
    if err != nil {
        fmt.Println(err)
        return errors.New("Incorrect path!")
    }
    defer inputFile.Close()

    scanner := bufio.NewScanner(inputFile)
    var rawData []string

    for scanner.Scan() {
        records := strings.Fields(scanner.Text())
        rawData = append(rawData, records...) 
    }

    // Check if the given file is actually 12th result file or not
    if strings.Join(rawData[4:8], " ") != "SENIOR SCHOOL CERTIFICATE EXAMINATION" {
        fmt.Println("It doesn't looks like a valid result file. Make sure the path is correct.")
        return errors.New("Not a valid 12th result file.")
    }

    parsedData := parser(rawData)

    outputFile, _ := os.Create("class_12th_result.csv")
    defer outputFile.Close()

    writer := csv.NewWriter(outputFile)
    
    err = writer.WriteAll(parsedData)
    if err != nil {
        fmt.Println(err)
        return errors.New("Error occurred while writing to .csv")
    } else {
        return nil
    }
}

func parser(rawData []string) [][]string {
    rollRegex, _ := regexp.Compile("^\\d{8}$")
    nameRegex, _ := regexp.Compile("([A-Z])\\w+")

    subjectCodes := map[string]string{
        "027": "HISTORY",
        "028": "POLITICAL SCIENCE",
        "029": "GEOGRAPHY",
        "030": "ECONOMICS",
        "041": "MATHEMATICS",
        "042": "PHYSICS",
        "043": "CHEMISTRY",
        "044": "BIOLOGY",
        "048": "PHYSICAL EDUCATION",
        "049": "PAINTING",
        "054": "BUSINESS STUDIES",
        "055": "ACCOUNTANCY",
        "083": "COMPUTER SCIENCE",
        "301": "ENGLISH CORE",
        "302": "HINDI CORE",
        "806": "TOURISM",
        "809": "FOOD PRODUCTION",
    }

    parsedData := [][]string{
        {
            "ROLL", "GENDER", "NAME", 
            "SUBJECT CODE 1", "SUBJECT NAME", "MARKS OBTAINED", "GRADE", 
            "SUBJECT CODE 2", "SUBJECT NAME", "MARKS OBTAINED", "GRADE", 
            "SUBJECT CODE 3", "SUBJECT NAME", "MARKS OBTAINED", "GRADE", 
            "SUBJECT CODE 4", "SUBJECT NAME", "MARKS OBTAINED", "GRADE", 
            "SUBJECT CODE 5", "SUBJECT NAME", "MARKS OBTAINED", "GRADE", 
            "SUBJECT CODE 6", "SUBJECT NAME", "MARKS OBTAINED", "GRADE",
            "TOTAL", "PERCENTAGE", "RESULT",
        },
    }

    for i, _ := range rawData {
        if rollRegex.MatchString(rawData[i]) {
            // Parse individual student's data 

            roll := rawData[i]
            gender := rawData[i + 1]

            var name string
            switch {
            // Extremely rare but if a student has five word name
            case nameRegex.MatchString(rawData[i + 6]):
                name = strings.Join(rawData[i + 2: i + 7], " ")
                i += 3

            // Rare but if a student has four word name
            case nameRegex.MatchString(rawData[i + 5]):
                name = strings.Join(rawData[i + 2: i + 6], " ")
                i += 2

            // If a student has three word name
            case nameRegex.MatchString(rawData[i + 4]):
                name = strings.Join(rawData[i + 2: i + 5], " ")
                i++

            // If a student has only one word name
            case !nameRegex.MatchString(rawData[i + 3]):
                name = rawData[i + 2]
                i--
            
            // Default two word names
            default:
                name = strings.Join(rawData[i + 2: i + 4], " ")
            }

            // Skip R.L or other results which doesn't have any data to it
            result := rawData[i + 13]
            if result != "PASS" && result != "COMP" {
                continue
            }

            var studentSubjects [][]string
            for _, j := range rawData[i + 4: i + 10] {
                studentSubjects = append(studentSubjects, []string{j, subjectCodes[j]})
            }

            var marks [][]string
            for j := 14; j < 25; j += 2 {
                marks = append(marks, []string{rawData[i + j], rawData[i + j + 1]})
            }

            var total int
            for _, j := range marks {
                num, _ := strconv.Atoi(j[0])
                total += num
            }
            additionalSubjectMarks, _ := strconv.Atoi(marks[5][0])
            percentage := (total - additionalSubjectMarks) / 5

            // Convert everything back to string
            strTotal := strconv.Itoa(total)
            strPercentage := strconv.Itoa(percentage)

            studentData := []string{roll, gender, name}
            for j := 0; j < 6; j++ {
                tempArray := []string{studentSubjects[j][0], studentSubjects[j][1], marks[j][0], marks[j][1]}
                studentData = append(studentData, tempArray...)
            }
            tempArray := []string{strTotal, strPercentage, result}
            studentData = append(studentData, tempArray...)

            parsedData = append(parsedData, studentData)
        }
    }
    return parsedData
}

File: cbse_class_12_txt_to_csv.py
# A luciferchase original
# https://gist.github.com/luciferchase/ede7a276f61a338d5c371eca945164fe

import csv
import re

roll_regex = re.compile("^\\d{8}$")
name_regex = re.compile("([A-Z])\\w+")

# Edit this to the path to your result file
path = ""

with open(path, "r") as f:
    raw_data = f.read()
    raw_data = raw_data.split()

subject_codes = {
    "027": "HISTORY",
    "028": "POLITICAL SCIENCE",
    "029": "GEOGRAPHY",
    "030": "ECONOMICS",
    "041": "MATHEMATICS",
    "042": "PHYSICS",
    "043": "CHEMISTRY",
    "044": "BIOLOGY",
    "048": "PHYSICAL EDUCATION",
    "049": "PAINTING",
    "054": "BUSINESS STUDIES",
    "055": "ACCOUNTANCY",
    "083": "COMPUTER SCIENCE",
    "301": "ENGLISH CORE",
    "302": "HINDI CORE",
    "806": "TOURISM",
    "809": "FOOD PRODUCTION"
}

parsed_data = [
    [
        "ROLL", "GENDER", "NAME", 
        "SUBJECT CODE 1", "SUBJECT NAME", "MARKS OBTAINED", "GRADE", 
        "SUBJECT CODE 2", "SUBJECT NAME", "MARKS OBTAINED", "GRADE", 
        "SUBJECT CODE 3", "SUBJECT NAME", "MARKS OBTAINED", "GRADE", 
        "SUBJECT CODE 4", "SUBJECT NAME", "MARKS OBTAINED", "GRADE", 
        "SUBJECT CODE 5", "SUBJECT NAME", "MARKS OBTAINED", "GRADE", 
        "SUBJECT CODE 6", "SUBJECT NAME", "MARKS OBTAINED", "GRADE",
        "TOTAL", "PERCENTAGE", "RESULT"
    ]
]

for i in range(len(raw_data)):
    # Filter anything which looks like a roll number
    if roll_regex.search(raw_data[i]):
        # Parse individual student's data

        roll, gender = raw_data[i: i + 2]

        # Extremely rare but if a student has five word name
        if (name_regex.search(raw_data[i + 6])):
            name = " ".join(raw_data[i + 2: i + 7])
            i += 3
        # Rare but if a student has four word name
        elif name_regex.search(raw_data[i + 5]):
            name = " ".join(raw_data[i + 2: i + 6])
            i += 2
        # If a student has three word name
        elif name_regex.search(raw_data[i + 4]):
            name = " ".join(raw_data[i + 2: i + 5])
            i += 1     
        # If a student has only one word name
        elif not name_regex.search(raw_data[i + 3]):
            name = raw_data[i + 2]
            i -= 1
        # Default two word names
        else:
            name = " ".join(raw_data[i + 2: i + 4])
        
        result = raw_data[i + 13]
        if (result not in ["PASS", "COMP"]):
            continue

        student_subjects = [[j, subject_codes[j]] for j in raw_data[i + 4: i + 10]]
        marks = [[raw_data[i + j], raw_data[i + j + 1]] for j in range(14, 25, 2)]
        int_marks = [int(i[0]) for i in marks]

        total = str(sum(int_marks))
        percentage = str((int(total) - int(raw_data[i + 24])) / 5)[:4]
        # If percentage is of the form XX.X add 0 to the last
        if (len(percentage) == 3):
            percentage += "0"

        student_data = [roll, gender, name]
        for j in range(6):
            student_data.extend([student_subjects[j][0], student_subjects[j][1], marks[j][0], marks[j][1]])
        student_data.extend([total, percentage, result])
        parsed_data.append(student_data)

with open("class_12th_result.csv", "w", newline = "\n") as f:
    writer = csv.writer(f)
    writer.writerows(parsed_data)